### PR TITLE
Catalog2AzureSearch changes for Search-by-TFM

### PR DESCRIPTION
**Pull request by advay26**
Addresses https://github.com/NuGet/Engineering/issues/4576

This change adds framework data to the Search Index through the Catalog2AzureSearch job:
* Read packageEntries (list of package's files) from the Catalog leaves
* Calculate a package's asset frameworks using catalog leaf properties (packageId, packageTypes, packageEntries)
* Populate 'Tfms' and 'Frameworks' index fields
* Add tests

### CODE REVIEW

File: src/NuGet.Services.AzureSearch/SearchDocumentBuilder.cs [hunk 1]
```diff
@@ -413,5 +440,29 @@ private static string[] ParseFrameworkGenerations(ICollection<NuGetFramework> tf
 
             return frameworks.ToArray();
         }
+
+        private static IEnumerable<NuGetFramework> GetSupportedFrameworks(PackageDetailsCatalogLeaf leaf)
+        {
+            string[] files = leaf.PackageEntries.Count == 0
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchDocumentBuilder.cs:446]: Could this package entries property be null? Perhaps for very old catalog leaves?

File: tests/NuGet.Services.AzureSearch.Tests/SearchDocumentBuilderFacts.cs [hunk 2]
```diff
@@ -563,6 +567,87 @@ public void SetsIconUrlToFlatContainerWhenPackageHasIconFileAndNoIconUrl()
...
                 Assert.Equal(Data.FlatContainerIconUrl, document.IconUrl);
             }
+
+            [Theory]
+            [MemberData(nameof(TargetFrameworkCases))]
+            public void AddsFrameworksAndTfmsFromCatalogLeaf(List<string> supportedFrameworks, List<string> expectedTfms, List<string> expectedFrameworks)
+            {
+                // arrange
+                var leaf = Data.Leaf;
+                leaf.PackageEntries = supportedFrameworks
```
Comment (joelverhagen) [inline tests/NuGet.Services.AzureSearch.Tests/SearchDocumentBuilderFacts.cs:577]: Consider baking a valid for PackageEntries property onto Data.Leaf by default. Ideally Data.Leaf has a good happy path value for all test cases and is only overridden for edge cases.

File: tests/NuGet.Services.AzureSearch.Tests/SearchDocumentBuilderFacts.cs [hunk 3]
```diff
@@ -563,6 +567,87 @@ public void SetsIconUrlToFlatContainerWhenPackageHasIconFileAndNoIconUrl()
...
+                    Data.Versions,
+                    isLatestStable: false,
+                    isLatest: true,
+                    normalizedVersion: Data.NormalizedVersion,
+                    fullVersion: Data.FullVersion,
+                    leaf: leaf,
+                    owners: Data.Owners);
+
+                // assert
+                Assert.True(document.Tfms.Length == expectedTfms.Count);
```
Comment (joelverhagen) [inline tests/NuGet.Services.AzureSearch.Tests/SearchDocumentBuilderFacts.cs:597]: There is an Assert.Equals which provides a better failure message.

File: tests/NuGet.Services.AzureSearch.Tests/SearchDocumentBuilderFacts.cs [hunk 4]
```diff
@@ -563,6 +567,87 @@ public void SetsIconUrlToFlatContainerWhenPackageHasIconFileAndNoIconUrl()
...
+            [Theory]
+            [MemberData(nameof(AdditionalCatalogTFMCases))]
+            public void CalculatesAssetFrameworksFromPackageEntriesAndPackageTypes(bool isTools, List<string> files, List<string> expectedTfms, List<string> expectedFrameworks)
+            {
+                // arrange
+                var leaf = Data.Leaf;
+                leaf.PackageEntries = files
+                                        .Select(f => new NuGet.Protocol.Catalog.PackageEntry { FullName = f })
+                                        .ToList();
+                if (isTools)
```
Comment (joelverhagen) [inline tests/NuGet.Services.AzureSearch.Tests/SearchDocumentBuilderFacts.cs:619]: Consider splitting out an if condition in your test to a separate test. From the philosophical level we call unit tests as facts and when the fact itself forks maybe it should be a separate fact. I'm not sure if that makes much sense, but having logic splits in tests can lead to confusing or convoluted tests to maintain.

File: tests/NuGet.Services.AzureSearch.Tests/SearchDocumentBuilderFacts.cs (hunk 2 above)
Comment (joelverhagen) [reply tests/NuGet.Services.AzureSearch.Tests/SearchDocumentBuilderFacts.cs:577]: Wait I see you've updated V3Data.cs with a default value for this property (I think). Is this override necessary?

**Review by joelverhagen — CHANGES_REQUESTED** [review]
Just some small comments, good overall

File: tests/NuGet.Services.AzureSearch.Tests/SearchDocumentBuilderFacts.cs (hunk 4 above)
Comment (advay26) [reply tests/NuGet.Services.AzureSearch.Tests/SearchDocumentBuilderFacts.cs]: I modified the test cases so that we don't need this if condition. We can just pick up the packageTypes from the test inputs directly.

File: tests/NuGet.Services.AzureSearch.Tests/SearchDocumentBuilderFacts.cs (hunk 2 above)
Comment (advay26) [reply tests/NuGet.Services.AzureSearch.Tests/SearchDocumentBuilderFacts.cs:577]: > Wait I see you've updated V3Data.cs with a default value for this property (I think).

I did do this. I wanted the other tests to reflect the Search-by-TFM changes too, so I modified the leaf and updated the  expected return values for the Frameworks and Tfms fields. https://github.com/NuGet/NuGet.Jobs/blob/c8b01a15eecf76bc937abc562655f23192171a98/tests/NuGet.Services.AzureSearch.Tests/SearchDocumentBuilderFacts.cs#L395-L400

File: src/NuGet.Services.AzureSearch/SearchDocumentBuilder.cs (hunk 1 above)
Comment (advay26) [reply src/NuGet.Services.AzureSearch/SearchDocumentBuilder.cs]: > Could this package entries property be null? Perhaps for very old catalog leaves?

And yes, packageEntries is null for leaves from before ~2016

**Review by joelverhagen — APPROVED** [review]

**Review by dannyjdev — APPROVED** [review]
LGTM, :shipit:
